Implement Income Details Graphic

sandrina-p opened the pull request:
Related to #556, implemented the missing graphic related to group pledging status.

Now, that it's connected with real data, there were some scenarios that I want to make sure they are okay visually:

1 - When nobody needs income, the group goal is 0$, so there's no blue bar and the surplus is always 100%:
<img width="929" alt="image" src="https://user-images.githubusercontent.com/14869087/69893658-c42c6300-130c-11ea-87dc-dea1b80621d8.png">

2 - a scenario where all pledge is your full income, nothing weird, but just to check. A reminder: the group goal is the total needed pledge. in this case $100
<img width="934" alt="image" src="https://user-images.githubusercontent.com/14869087/69893759-233ea780-130e-11ea-92a8-0d1efedfb85e.png">

3 - a scenario where surplus + your income fulfill 100% the graphic. nothing wrong, just to check:
<img width="944" alt="image" src="https://user-images.githubusercontent.com/14869087/69893858-7533fd00-130f-11ea-811f-e8935d1115f6.png">

4 - Now things get interesting: (🐛 bug ahead!) The surplus is bigger than the group goal ($30). Your mincome (orange) conflicts with the surplus. As it is now, the surplus on the graphic has a maximum percentage of 100%, this is, the group goal. Your mincome has the same value - 100% - as in the first scenario. Which one should appear first? Should they appear together? If yes, should we split their space accordingly (in this case 70% to surplus and 30% to income?)
<img width="929" alt="image" src="https://user-images.githubusercontent.com/14869087/69893802-a3650d00-130e-11ea-8aaa-62a1f5f07002.png">

Well... it's easier to understand all of this if you try the graphic by yourself. Just create a group with 2 members and that's all you need to see the previous scenarios.

### Misc stuff:
- Add prop `fullscreen` to the modal component and improve its UI (paddings around + missing ESC on the close button). Note: To keep PR small, I only applied this to the Income Details modal. In a next PR it should be applied to all the other fullscreen modals and [address the issue with general banner](https://github.com/okTurtles/group-income-simple/pull/779#issuecomment-559900573)
- Move from `group-proposals.spec.js` the tests related with payment to a new cypress spec `group-payments`. It's a raw copy-paste with a few more assertions related to the Income Details Graphic.

sandrina-p committed:
Implement Income Details Graphic

sandrina-p committed:
implement graph UI

sandrina-p committed:
optimize cypress tests organization

sandrina-p committed:
incomeDetails: connect real data to graphic +  write cypress tests

sandrina-p committed:
IncomeDetails - Continue graph refinement

sandrina-p committed:
Modal - design tweaks on fullscreen variant

sandrina-p commented on .Gruntfile.babel.js:
```diff
@@ -337,7 +337,7 @@ module.exports = (grunt) => {
           recurse: true
         }),
-        eslint({ throwOnError: true, throwOnWarning: true }),
+        eslint({ throwOnError: true, throwOnWarning: true, fix: true }),
```
NOTE: 👮‍♀ I won't forget to delete this before merging

taoeffect replied:
seems about time to remove :)

sandrina-p commented on frontend/views/containers/contributions/IncomeDetails.vue:
```diff
@@ -1,79 +1,76 @@
-      i18n.is-title-2.c-title(tag='h2') Income Details
+modal-base-template(ref='modal' :fullscreen='true')
+  .c-content
+    i18n.is-title-2.c-title(tag='h2') Income Details
```
NOTE: I replaced this extra wrapper with the new modal prop fullscreen. The indentation changed! That's why this huge diff..

Just review the `group-pledges-graph` component, that is the real change.

pieer replied:
There is already 2 kind of modals, one full screen and one not. Is this a 3rd kind?

sandrina-p replied:
As far as I understand, we don't have a "real" fullscreen modal. To implement a fullscreen modal we have been following different approaches:
1- create a wrapper child of `modal-base-template` and style it, adding the dimensions and the missing `background-color: $general_2`. Examples: this modal, PayGroupHistory, etc...
2 - Style `.modal` "indirectly". I say indirectly because it's not explicit that it's a class inside `modal-base-template`, for example `CreateGroup.vue` or `Join.vue`. There we also manually add the `background-color`. 
3 - Add that same bg to some elements inside the modal, for example, `SettingsWrapper` (the user settings modal), where we add the bg to its sidebar and its main.

And, no matter the approach, all of these modals are incorrectly styled because the paddings and the button style (having ESC) are wrong. 
For those reasons, I decided to explicit add a `fullscreen` prop to `modal-base-template` that deal with that for us. No need to add the bg color. It comes already centered aligned and with the correct paddings and close button styled correctly.

Does this make sense?

sandrina-p committed:
Assert group pledging graphic

sandrina-p commented on frontend/views/containers/GroupPledgesGraph.vue:
```diff
@@ -1,218 +1,195 @@
+        const membersNeedingPledges = this.groupMembersCount - othersPledgesCount
+        const avgNeededPledges = neededPledges / membersNeedingPledges
+        const avgMembersNeedingPledges = groupGoal / membersNeedingPledges
+        ourIncomeToReceive = +(ourIncomeNeeded - (ourIncomeNeeded * avgNeededPledges / avgMembersNeedingPledges)).toFixed(0)
```
Question: @taoeffect can u suggest a clean/easy way of using `IncomeDistribution` here without repeating the code in the getter `groupIncomeDistribution`?

taoeffect replied:
I'm taking it easy today, but for when I get back, can you let me know what you mean by "using `IncomeDistribution`"? I don't see that defined anywhere.

sandrina-p replied:
IncomeDistribution, I refer to the following file imported at `state.js`, used to create the getter `groupIncomeDistribution`:

> import incomeDistribution from '~/frontend/utils/distribution/mincome-proportional.js'

taoeffect replied:
OK thanks! I'll have a close look at what you're doing here when after the weekend (or perhaps tomorrow, we'll see)

taoeffect replied:
OK, yes, so here's what needs to be done:

1. Up above in the loop starting at line 82, convert that to a for/in loop that looks similar to the loop in `groupIncomeDistribution` in order to fill up an array called `newIncomeDistribution` (just like the `currentIncomeDistribution` inside of `groupIncomeDistribution`)
2. Outside of the loop, perhaps in here, push again to newIncomeDistribution `{ name: username, amount: ourIncomeAmount }`
3. In here, do something like `ourIncomeToReceive = incomeDistribution(newIncomeDistribution, mincome).filter(i => i.to === username).reduce((acc, cur) => cur.amount + acc, 0)` (double-check me there) to generate `ourIncomeToReceive`

mmbotelho commented:
> 4 - Now things get interesting: (🐛 bug ahead!) The surplus is bigger than the group goal ($30). Your mincome (orange) conflicts with the surplus. As it is now, the surplus on the graphic has a maximum percentage of 100%, this is, the group goal. Your mincome has the same value - 100% - as in the first scenario. Which one should appear first? Should they appear together? If yes, should we split their space accordingly (in this case 70% to surplus and 30% to income?)

The income to receive should be displayed first - it's more relevant to the user.

sandrina-p commented:
> The income to receive should be displayed first - it's more relevant to the user.

Okay, it makes sense. What about the next question: "Should they appear together?"

mmbotelho commented:
@sandrina-p from what I can see, the current design does not handle these edge cases very well... For our sanity, let's just show either one or the other (income to receive has priority over the surplus). 

We can come back later and iterate on the design and implementation 😄what do you think?

sandrina-p commented:
> For our sanity, let's just show either one or the other (income to receive has priority over the surplus).

So, we should show only income to receive, even if there's space for surplus as well? (scenario 3)

mmbotelho reviewed (changes requested):
Looking good @sandrina-p ! Here's some things I found:
<img width="683" alt="Captura de ecrã 2019-11-30, às 18 28 48" src="https://user-images.githubusercontent.com/22004566/69903937-93960900-13a0-11ea-9871-b1dfe8360917.png">
<img width="633" alt="Captura de ecrã 2019-11-30, às 18 24 58" src="https://user-images.githubusercontent.com/22004566/69903939-93960900-13a0-11ea-919e-dab217b313bf.png">
<img width="1038" alt="Captura de ecrã 2019-11-30, às 18 27 14" src="https://user-images.githubusercontent.com/22004566/69903940-93960900-13a0-11ea-8b4d-ea9d2286801b.png">
<img width="488" alt="Captura de ecrã 2019-11-30, às 18 36 34" src="https://user-images.githubusercontent.com/22004566/69903976-00110800-13a1-11ea-97e1-06f14bd07f78.png">


-----


Probably not related with this PR: When a user has a $0 income, I get this "false" on the contributions page:
<img width="886" alt="Captura de ecrã 2019-11-30, às 18 32 53" src="https://user-images.githubusercontent.com/22004566/69903967-d952d180-13a0-11ea-9065-68180a579965.png">

<img width="486" alt="Captura de ecrã 2019-11-30, às 18 35 10" src="https://user-images.githubusercontent.com/22004566/69903975-f7b8cd00-13a0-11ea-8ccb-1544922d74de.png">

taoeffect commented on frontend/views/containers/GroupPledgesGraph.vue:
```diff
@@ -1,218 +1,195 @@
+      let othersPledgesAmount = 0
+      let othersPledgesCount = 0
+
+      Object.keys(this.groupProfiles).forEach(username => {
```
I recommend changing this to a `for/in` loop, as they're faster and to match the style of code in `groupIncomeDistribution`, because this loop will be performing a similar function.

Line 85 will need to be updated (you can just wrap the lines below it in a `if (contractID !== this.ourContractId) { .. }`)

taoeffect commented on frontend/model/state.js:
```diff
@@ -243,6 +243,9 @@ const getters = {
   ourUsername (state) {
     return state.loggedIn && state.loggedIn.username
   },
+  ourContractId (state) {
```
"ContractId" could refer to any number of contracts, it's important to specify which contractID this is returning; `ourContractId` => `ourIdentityContractId`

taoeffect commented on frontend/views/components/Tooltip.vue:
```diff
@@ -113,7 +113,7 @@ export default {
 @import "@assets/style/_variables.scss";
 
-.c-wrapper {
+.c-twrapper {
```
What is the reason for this change?

pieer replied:
It should not affect the rest because of the scope, I would keep the c-wrapper. If there is a generic css call c-wrapper, that's a mistake

sandrina-p replied:
I should have explained this... actually renaming `c-wrapper` to `c-twrapper` or `c-gwrapper` (at groupPledgesGraph) has a reson, although I don't like it very much. Maybe @pieer you know a better solution. So, let me explain, by starting to rollback the class name to `c-wrapper` as before:

![vue-bug](https://user-images.githubusercontent.com/14869087/70040049-c6afe680-15b2-11ea-90d8-86a3eb40631b.png)

See those classes in the DOM `c-wrapper`, one for the graphic (blue outline) and other for the tooltip (red outline). For some reason, the `data-v-59ba5d0f` attribute is added also to the tooltip, so the tooltip will have not only its styles (orange styles) but also the styles applied to the graphic wrapper (green styles), breaking the layout.

After some experiments, I realized this only happens when applied to the component top-level element (the wrapper of everything inside that component).

This is really annoying because it made me realize that CSS in Vue isn't truly scoped. So for that reason using `c-wrapper` or `c-container` as the top class of every component can potentially lead to unexpected bugs as this one. So, I believe we should find a little more "unique" names, at least for those top-level elements. That's why I wrote `c-gwrapper` and `c-twrapper`.

taoeffect replied:
Wow, @sandrina-p ... is there a way you can add a shortened version of this explanation to the Style Guidelines CSS section? I'm still having some difficulty understanding it, but it's a bit clearer to me now. Maybe link to this comment as well.

Also, yeah, if @pieer you have any ideas on how to avoid this, please do let us know!

taoeffect commented on frontend/views/containers/GroupPledgesGraph.vue:
```diff
@@ -1,218 +1,195 @@
-      const userIncomeToReceive = userIncomeNeeded ? (userIncomeNeeded - (userIncomeNeeded * avgNeededPledges / avgMembersNeedingPledges)).toFixed(2) : 0
+      const mincome = this.groupSettings.mincomeAmount
+      const ourPledgeAmount = this.type === 'pledgeAmount' && this.amount >= 0 && this.amount
+      const ourIncomeAmount = this.type === 'incomeAmount' && this.amount < mincome && this.amount
```
I think the checks in the middle (`this.amount >= 0` and `this.amount < mincome`) aren't needed, because vuelidate should perform those checks itself.

sandrina-p replied:
Those are needed, otherwise, the graphic would display strange values such as negative values.

taoeffect commented on frontend/views/containers/GroupPledgesGraph.vue:
```diff
@@ -1,218 +1,195 @@
+      const mincome = this.groupSettings.mincomeAmount
+      const ourPledgeAmount = this.type === 'pledgeAmount' && this.amount >= 0 && this.amount
+      const ourIncomeAmount = this.type === 'incomeAmount' && this.amount < mincome && this.amount
+      const doWeNeedPledge = typeof ourIncomeAmount === 'number'
```
Think we can avoid performing redundant computation with `const doWeNeedPledge = this.type === 'incomeAmount'`, and then replacing redundant checks like `this.type === 'incomeAmount'` (e.g. the ones on line 76 and 75) with `doWeNeedPledge` and `!doWeNeedPledge` (since they are mutually exclusive)

taoeffect commented on frontend/views/containers/GroupPledgesGraph.vue:
```diff
@@ -1,218 +1,195 @@
+      const pledgeTotal = othersPledgesAmount + ourPledgeAmount
+      const groupGoal = othersIncomeNeeded + ourIncomeNeeded
+      const neededPledges = groupGoal - pledgeTotal
+      const surplus = pledgeTotal - othersIncomeNeeded - ourIncomeNeeded
```
```js
const neededPledges = Math.max(0, groupGoal - pledgeTotal)
const surplus = Math.max(0, pledgeTotal - othersIncomeNeeded - ourIncomeNeeded)
```

Then can remove the [branches](https://en.wikipedia.org/wiki/Branch_(computer_science)) on lines 121 and 122

taoeffect commented on frontend/views/containers/GroupPledgesGraph.vue:
```diff
@@ -1,218 +1,195 @@
+      const neededPledges = groupGoal - pledgeTotal
+      const surplus = pledgeTotal - othersIncomeNeeded - ourIncomeNeeded
+
+      let ourIncomeToReceive
```
This can be initialized to `ourIncomeNeeded` so that lines 110-112 can be removed

taoeffect commented on frontend/views/containers/GroupPledgesGraph.vue:
```diff
@@ -1,218 +1,195 @@
+      const surplus = pledgeTotal - othersIncomeNeeded - ourIncomeNeeded
+
+      let ourIncomeToReceive
+      if (neededPledges > 0) {
```
This would then be:

```js
if (doWeNeedPledge && neededPledges > 0) {
```

sandrina-p replied:
there's no need to doWeNeedPledge. `neededPledges > 0` is enough

taoeffect replied:
`neededPledges > 0` can be true even when `doWeNeedPledge` is false, and when `doWeNeedPledge` is false, we don't need to run this calculation, right?

taoeffect reviewed (changes requested):
See comments + merge conflicts!

pieer commented on frontend/assets/style/_reset.scss:
```diff
@@ -34,7 +34,7 @@ h3,
 h5,
 h6 {
-  font-weight: bold;
+  font-weight: normal;
```
Wow look like a big change for that pull request.. but I guess you've tested everywhere where there is a title that it look like the design, isn't it?

sandrina-p replied:
Yes, I did exactly that on the last PR where we changed `is-X` to `is-title-X`, so now, *every* title must have the `is-title-x` class, so there's no need to add `font-weight: bold` to heading elements. When we need to use a heading element without looking like a title, we don't need to override its style anymore.

pieer commented on frontend/assets/style/_typography.scss:
```diff
@@ -37,8 +37,7 @@ body {
   font-size: $size-4; // 14px
-  line-height: 1.3125rem; // 21px
-  letter-spacing: 0.1px;
+  line-height: 1.5; // 21/14
```
Same here ;)

sandrina-p replied:
`line-height` 1.3125rem is exactly the same as 1.5, so it was just a matter of readability.
`letter-spacing` was removed because it isn't needed - I confirmed with @mmbotelho.

pieer commented on frontend/assets/style/_typography.scss:
```diff
@@ -153,6 +152,14 @@ small {
     background-color: var(--#{$name}_2);
   }
+
+  .has-background-#{$name}-solid {
```
Good call

pieer commented on frontend/views/components/Graphs/PieChart.vue:
```diff
@@ -1,53 +1,43 @@
 <template lang='pug'>
 .c-container
-  svg.c-piechart(viewbox='-1 -1 2 2' :style='{ height: size }')
+  svg.c-piechart(viewBox='-1 -1 2 2' aria-hidden='hidden')
```
Interesting viewbox. Could you had a comment please?

sandrina-p replied:
Yes, sure, there's a comment below with a URL explaining more about how this piechart was made:
// Learn more about SVG & PieCharts
// -> https://hackernoon.com/a-simple-pie-chart-in-svg-dbdd653b6936

pieer commented on frontend/views/components/Graphs/PieChart.vue:
```diff
@@ -166,4 +150,18 @@ export default {
   line-height: 1.2;
 }
+
+// Global Classes for SVG fills
```
Why did you move it from utilities? I could use the same class for my graph actually

sandrina-p replied:
You are right! Moved back to utilities and with better names.

pieer reviewed (commented):
Good call job on refactoring group proposal test.
I'm not so sure about the full screen modal that look like overlapping on the other one. But I guess you must have you reason (could you explain quickly?)
I fill that changing the basic font property (not bold) for all titles is surprising at this time of the project, but again I trust your judgement. 
In general, I think that updating something that affect all the website for a pull request with the title implement graphic is not really appropriate. But above all, it's good that you improve the full website when you can... so I let you judge.

sandrina-p committed:
review all fullscreen modals (5) and update them with the new prop fullscreen

sandrina-p committed:
design tweaks based on review

sandrina-p committed:
Remove useless letter-spacing 0.1px

mmbotelho commented:
> So, we should show only income to receive, even if there's space for surplus as well? (scenario 3)

Yes, or else it's very confusing

sandrina-p committed:
Changes based on CR

sandrina-p commented:
@greg, @pieer addressed all the issues you've pointed. Please take a look again.

@mmbotelho same for you, even those not directly related with this PR.

mmbotelho reviewed (changes requested):
Only missing this:
<img width="493" alt="Captura de ecrã 2019-12-03, às 14 00 31" src="https://user-images.githubusercontent.com/22004566/70053488-d8ef4c00-15d5-11ea-8f52-73be11383b70.png">

sandrina-p committed:
last ui tweak

sandrina-p commented:
@mmbotelho done!

mmbotelho reviewed (dismissed):
Perfect! Thank you @sandrina-p 😄

taoeffect commented on frontend/views/containers/GroupPledgesGraph.vue:
```diff
@@ -1,218 +1,204 @@
+      'ourUsername'
+    ]),
+    doWePledge () {
+      return this.type === 'pledgeAmount'
```
Note: since this function is only called from within `graphData()`, it is better to keep this as a variable in that function. A function call can add significant overhead, and the `graphData()` function has significant computation in it. So anything unnecessary should be removed. So I recommend defining this as a `doWeNeedPledge` const inside `graphData()`

taoeffect commented on frontend/views/containers/GroupPledgesGraph.vue:
```diff
@@ -1,218 +1,204 @@
+      // avoid breaking the graph when perc is bigger than 1,
+      if (perc > 1) return 1
+      if (perc < 0) return 0
+      return perc
```
can save some lines:

```
return Math.max(1, Math.min(0, perc))
```

sandrina-p replied:
This is wrong @taoeffect , that's what causing the bug you are facing (missing the white slice)

taoeffect replied:
Oops, it should have been:

```js
return Math.min(1, Math.max(0, perc))
```

sandrina-p replied:
It makes sense now (and it works!) changed

taoeffect commented on frontend/views/containers/contributions/IncomeDetails.vue:
```diff
@@ -197,7 +194,7 @@ export default {
         [L('This field is required')]: required,
         [L("It seems your income is not lower than the group's mincome.")]: function (value) {
```
Boyscoutting:

`"It seems your income is not lower than the group's mincome."` => `"Income must be lower than group mincome"`

mmbotelho replied:
Can we write a complete sentence so it doesn't sound so robotic? Maybe `Your income must be lower than the group mincome.`

taoeffect replied:
Agreed, that sound even better :)

taoeffect commented on frontend/views/containers/contributions/IncomeDetails.vue:
```diff
@@ -197,7 +194,7 @@ export default {
-            return value < this.groupSettings.mincomeAmount - 1
+            return value < this.groupSettings.mincomeAmount
           }
           return true
```
save some lines:

```
return !this.needsIncome || value < this.groupSettings.mincomeAmount
```

taoeffect reviewed (changes requested):
Some minor suggestions + don't forget to update .Gruntfile.babel.js :)

sandrina-p committed:
minor improvements and rollback gruntfile

sandrina-p commented:
@greg, changes made and branch updated with master

sandrina-p committed:
Update tests based on last currencies change to display decimals

sandrina-p committed:
minor improvement on ourIncomeToReceive

sandrina-p committed:
improve error message

sandrina-p committed:
fix typo

taoeffect reviewed (changes requested):
## Setup

1. Created a group with Person A and Person B
2. Person A - set their monthly income to $600

### Issue 1 - Incorrect center value number and slices

<img width="938" alt="Screen Shot 2019-12-17 at 9 11 37 AM" src="https://user-images.githubusercontent.com/138706/71018643-b9642300-20ad-11ea-900f-8dd4262b9d0a.png">

Here Person B has gone to add their income details, and they're told that the Group Goal is $1400. This is what's shown when neither option is picked, I guess because it's assumed that Person B needs income and has $0 income themselves.

We can see how the Group Goal actually *decreases* when they say their monthly income is $300:

<img width="936" alt="Screen Shot 2019-12-17 at 9 13 17 AM" src="https://user-images.githubusercontent.com/138706/71018763-f16b6600-20ad-11ea-9b06-ea20cf432f8e.png">

Instead, when no option is picked, the graph should show different numbers in the center, depending on whether or not there's enough pledges. In this case, there wasn't enough pledges to cover the amount needed, so it should show "$400 needed".

Now let's look at what happens if we have Person B decide to pledge $300:

<img width="931" alt="Screen Shot 2019-12-17 at 9 12 57 AM" src="https://user-images.githubusercontent.com/138706/71019030-7bb3ca00-20ae-11ea-80a2-cfc0526abfa0.png">

Here we see two issues:

1. The pie chart should show enough slices to illustrate the situation: that $300 is being pledged, that $400 was needed, and that $100 is still needed, which means we need to have at least 2 slices in the graph (or 3), but at the moment only one blue slice is shown.
2. The center value of the chart should change, to show how much is still needed, but here it stays fixed, saying that $400 is the "Group Goal". Instead it should change to show that `1000 - (600 + 300) = $100` is still needed to meet the mincome for the group

I stopped playing with the chart at this point, but this is the process I took, to see if the chart fully tells the story of what's going on in the group, depending on which radio button is selected, and what value is entered for either the pledge or the user's income.

### Issue 2 - Travis build still failing

Not sure if the latest failure is an intermittent one or not, but would be great if it can be figured out so that the Travis builds are consistent again.

sandrina-p committed:
Fix Typo with ! and add hasData

sandrina-p commented:
@greg, it's fixed! iI was a typo based on your suggestion, I added `doWePledge` and it should be `!doWePledge` so that’s why the data was wrong. Cypress failed for that same reason (because we have tests to cover those scenarios)

Also, I added `hasData` variable to exclude the user income from the group goal when the user didn't answer the question yet.

taoeffect reviewed (changes requested):
<img width="950" alt="Screen Shot 2019-12-17 at 10 07 38 AM" src="https://user-images.githubusercontent.com/138706/71022345-1dd6b080-20b5-11ea-876a-6e7aca08d5b5.png">

Looks like the white slice is still missing

sandrina-p committed:
undo decimalSlice proposed logic

sandrina-p commented:
@greg, the white slice is fixed. This [suggestion you made related to `decimalSlice()`](https://github.com/okTurtles/group-income-simple/pull/786#pullrequestreview-330227352) was the cause of the issue.

sandrina-p committed:
optmize decimalSlice logic

taoeffect reviewed (approved):
Approved!!! Great job @sandrina-p! The pie chart is so much fun to play with, it seems to work great in my testing!! 😄
